Add functions add/sub/rsub/mul/div that accept a scalar as first arg

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29238 Remove specific processing of declarations of add, sub, mul, div, rsub
* **#29602 Add functions add/sub/rsub/mul/div that accept a scalar as first arg**

Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.

xuhdev committed:
Add functions add/sub/rsub/mul/div that accept a scalar as first arg

[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

ailzhang reviewed (changes requested):
Hmmmm this is a lot of overloads...
I'd take a step back and do the design carefully first. 
There are two benefits of having this PR:
1. Currently in C++ frontend `torch::add(1, tensor)` doesn't compile. This PR will solve that. 
2. This PR combines the next in the stack together will remove the hack in `python_args_parser` where we convert a scalar directly to a tensor. 
My main concern about this PR is:
1. Number of overloads added and the future maintenance burden with it. 
cc: @ezyang @gchanan @yf225 to comment here. 
(Sorry I'm not really requesting changes but to make sure we don't proceed without a design.

ezyang commented:
I shouldn't make the call here, but my 2c is that for binary ops, it seems acceptable to me to have `(Tensor, Tensor)`, `(Tensor, Scalar)` and `(Scalar, Tensor)` overloads; it's a similar situation to how Python magic methods have `__add__` and `__radd__`. If we get ops with a lot more arguments, doing every permutations of scalars is a much harder sell, but fortunately, we don't support scalars for those anyway!

yf225 commented:
I think this is an important addition for Python/C++ API parity, otherwise there will be friction and ongoing maintenance overhead if the user has to translate Python code `torch.add(1, tensor)` to C++ code `torch::add(tensor, 1)`, as we have to think whether this is strictly equivalent to the Python version every time we look at the C++ code.

nairbv commented:
Is there some other approach we can take with this, even if it's not an immediate change in this PR but something to think about for refactoring longer-term?

I could imagine having functions be polymorphic in some way where we don't have to define any functions as accepting Scalar (other than in cases where they explicitly do not accept tensors). I'd think any function that can accept a tensor parameter should also be able to accept a scalar, and have it be interpreted as a zero-dim "wrapped number" tensor without extra handling or interfaces.

It's odd to me that we can torch.pow(tensor, scalar), torch.pow(scalar,tensor), but not torch.pow(scalar, scalar). If we rely on implementing every combination manually, we'll always have some gaps. I know it's probably uncommon to want to use torch.pow() on two scalars, but consistency helps keep user code generic across inputs. I imagine there are at least a few ways we could solve for this.

nairbv commented:
Also, I think these new interfaces would need to be handled in shape_analysis.cpp?

xuhdev commented:
@nairbv We might be able to use some templating with some helpers, but I think that might likely also require some other changes in how function declarations are generated...

ezyang commented:
I mean, we can talk all we want about how exactly the C++ API should be implemented, but at the end of the day, we still need to actually implement kernels for each of the overloads. This isn't evident from this PR because it's still using the old mechanism (`wrap_scalar_tensor`), but my understanding from performance discussion is that we don't want to be doing this if we want to speed up our scalar operations.

gchanan commented on test/test_torch.py:
```diff
@@ -9841,6 +9841,25 @@ def test_add(self, device):
                                lambda: torch.add(m1, m2, alpha=1.0))
 
+        # Scalar
+        for dtype in torch.testing.get_all_dtypes():
```
why are there python tests in this PR?  AFAIK this PR doesn't have any effect on python -- shouldn't these tests be on the the PR that changes the python binding and shouldn't we have C++ tests here?

xuhdev replied:
They will take effect after the stacked PR is merged

gchanan replied:
right, but they should be on the PR that makes the change, not on the prereq PR.

xuhdev replied:
I have moved the test code to the next PR.

gchanan commented on test/test_torch.py:
```diff
@@ -9841,6 +9841,25 @@ def test_add(self, device):
+        # Scalar
+        for dtype in torch.testing.get_all_dtypes():
+            m1 = torch.tensor([1, 2], dtype=dtype, device=device)
+            if dtype == torch.bool:
```
can you make the these use the generic test framework at the bottom so we can ensure XLA has these semantics?

xuhdev replied:
Are you referring to `test_*(self, device, dtype)` style test cases?

gchanan replied:
yes, that's what I mean.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3016,6 +3035,11 @@
   variants: function
   supports_named_tensor: True
 
+- func: rsub.Scalar2(Scalar other, Tensor self, Scalar alpha=1) -> Tensor
```
do we need rsub?  Why would I call this in C++?  Is it necessary for the python binding later?

xuhdev replied:
I don't think we need `rsub` (including the original ones), but we may need `rsub_`. I added merely because there are `rsub`s before. I can attempt to delete the existing `rsub`s if they are truly unecessary.

gchanan commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -164,6 +164,10 @@ Tensor add(const Tensor& self, Scalar other, Scalar alpha) {
 }
 
+Tensor add(Scalar other, const Tensor& self, Scalar alpha) {
+  return native::add(self, other, alpha);
```
this doesn't look right.

In the "normal" version, this op is:
`add(p1, p2,  alpha) -> p1 + p2*alpha`

but you are computing:
`add(p1, p2, alpha) -> p2 + p1*alpha`

though renaming all the variables makes this confusing (it doesn't matter in C++).

xuhdev replied:
Good point! I'll fix this and add a test.

gchanan commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -194,6 +206,10 @@ Tensor sub(const Tensor& self, Scalar other, Scalar alpha) {
 }
 
+Tensor sub(Scalar other, const Tensor& self, Scalar alpha) {
+  return native::sub(wrapped_scalar_tensor(other), self, alpha);
```
this is wrong, same as add.

xuhdev replied:
I think this is correct? It is different from `add`.

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

ezyang commented:
Removing myself from reviewer list. If there's something specific you want me to look at let me know.

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 8df7cb59:


* **3/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8df7cb59690cc129a717e155c5a121480a08d37f).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4285158) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87435287) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87435287))
<details>
<summary>
<code>Jan 16 23:29:52 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o</code>
</summary>

```
Jan 16 23:29:52 ../c10/core/MemoryFormat.h:32:75: warning: ‘c10::MemoryFormat c10::get_contiguous_memory_format()’ is deprecated [-Wdeprecated-declarations] 
Jan 16 23:29:52  #define LEGACY_CONTIGUOUS_MEMORY_FORMAT c10::get_contiguous_memory_format() 
Jan 16 23:29:52                                                                            ^ 
Jan 16 23:29:52 ../aten/src/ATen/native/Activation.cpp:681:39: note: in expansion of macro ‘LEGACY_CONTIGUOUS_MEMORY_FORMAT’ 
Jan 16 23:29:52    auto buffer = at::zeros_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT); 
Jan 16 23:29:52                                        ^ 
Jan 16 23:29:52 ../c10/core/MemoryFormat.h:34:36: note: declared here 
Jan 16 23:29:52  C10_DEPRECATED inline MemoryFormat get_contiguous_memory_format() { 
Jan 16 23:29:52                                     ^ 
Jan 16 23:29:52 [1660/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o 
Jan 16 23:29:52 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o  
l -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o -c ../aten/src/ATen/native/BinaryOps.cpp 
Jan 16 23:29:52 ../aten/src/ATen/native/BinaryOps.cpp: In function ‘at::Tensor& at::native::__ior__(at::Tensor&, c10::Scalar)’: 
Jan 16 23:29:52 ../aten/src/ATen/native/BinaryOps.cpp:350:61: error: a function-definition is not allowed here before ‘{’ token 
Jan 16 23:29:52  Tensor rsub(Scalar other, const Tensor& self, Scalar alpha) { 
Jan 16 23:29:52                                                              ^ 
Jan 16 23:29:52 ../aten/src/ATen/native/BinaryOps.cpp:354:82: error: a function-definition is not allowed here before ‘{’ token 
Jan 16 23:29:52  Tensor& bitwise_xor_out(Tensor& result, const Tensor& self, const Tensor& other) { 
Jan 16 23:29:52                                                                                   ^ 
Jan 16 23:29:52 ../aten/src/ATen/native/BinaryOps.cpp:361:61: error: a function-definition is not allowed here before ‘{’ token 
Jan 16 23:29:52  Tensor bitwise_xor(const Tensor& self, const Tensor& other) { 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4285171) pytorch_linux_xenial_py3_clang5_mobile_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87435631) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87435631))
<details>
<summary>
<code>Jan 16 23:30:32 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:572:1: error: expected '}' at end of input</code>
</summary>

```
Jan 16 23:30:32  Tensor& bitwise_xor_out(Tensor& result, const Tensor& self, const Tensor& other) { 
Jan 16 23:30:32                                                                                   ^ 
Jan 16 23:30:32 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:361:61: error: a function-definition is not allowed here before '{' token 
Jan 16 23:30:32  Tensor bitwise_xor(const Tensor& self, const Tensor& other) { 
Jan 16 23:30:32                                                              ^ 
Jan 16 23:30:32 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:572:1: error: expected '}' at end of input 
Jan 16 23:30:32  }  // namespace at 
Jan 16 23:30:32  ^ 
Jan 16 23:30:32 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp: At global scope: 
Jan 16 23:30:32 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:572:1: error: expected '}' at end of input 
Jan 16 23:30:32 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:572:1: error: expected '}' at end of input 
Jan 16 23:30:32 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o] Error 1 
Jan 16 23:30:32 caffe2/CMakeFiles/torch_cpu.dir/build.make:2296: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o' failed 
Jan 16 23:30:32 make[2]: *** Waiting for unfinished jobs.... 
Jan 16 23:30:38 In file included from /var/lib/jenkins/workspace/build_mobile/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 16 23:30:38                  from /var/lib/jenkins/workspace/aten/src/ATen/Tensor.h:11, 
Jan 16 23:30:38                  from /var/lib/jenkins/workspace/aten/src/ATen/Context.h:4, 
Jan 16 23:30:38                  from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:5, 
Jan 16 23:30:38                  from /var/lib/jenkins/workspace/aten/src/ATen/native/Distributions.cpp:1: 
Jan 16 23:30:38 /var/lib/jenkins/workspace/aten/src/ATen/native/Distributions.cpp: In function 'at::Tensor at::native::bernoulli(const at::Tensor&, at::Generator*)': 
Jan 16 23:30:38 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:32:46: warning: 'c10::MemoryFormat c10::get_contiguous_memory_format()' is deprecated [-Wdeprecated-declarations] 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

ezyang commented:
What's the status on this PR?

xuhdev commented:
@ezyang Still have trouble in understanding where things went wrong. Seems like this PR has touched a lot of spots where the assumption that the first arg is a tensor is entrenched.

ezyang commented:
OK, so you need help debugging it?

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev commented:
@ezyang  Yes. For example, for this error, I'm having a hard time figuring out the location of test code in `test_add_constant_cpu (test_jit.TestJitGeneratedAutogradCPU)`:

```
Jan 17 01:27:19 ======================================================================
Jan 17 01:27:19 ERROR [0.024s]: test_add_constant_cpu (test_jit.TestJitGeneratedAutogradCPU)
Jan 17 01:27:19 ----------------------------------------------------------------------
Jan 17 01:27:19 Traceback (most recent call last):
Jan 17 01:27:19   File "/var/lib/jenkins/workspace/test/common_device_type.py", line 180, in instantiated_test
Jan 17 01:27:19     return test(self, device_arg)
Jan 17 01:27:19   File "/var/lib/jenkins/workspace/test/test_jit.py", line 17436, in do_test
Jan 17 01:27:19     check(name)
Jan 17 01:27:19   File "/var/lib/jenkins/workspace/test/jit_utils.py", line 592, in wrapper
Jan 17 01:27:19     fn(*args, **kwargs)
Jan 17 01:27:19   File "/var/lib/jenkins/workspace/test/test_jit.py", line 17401, in check
Jan 17 01:27:19     check_types=check_types)
Jan 17 01:27:19   File "/var/lib/jenkins/workspace/test/test_jit.py", line 16972, in check_against_reference
Jan 17 01:27:19     allow_unused=allow_unused)
Jan 17 01:27:19   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 157, in grad
Jan 17 01:27:19     inputs, allow_unused)
Jan 17 01:27:19 RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
```

ezyang commented:
This is one of the metaprogrammed tests from the test suite. The best way to debug, IMO, is to edit the metaprogramming code, e.g., File "/var/lib/jenkins/workspace/test/jit_utils.py", line 592, in wrapper, while running only the test you care about in question. When that doesn't give me enough information I tend to try to reproduce the error in a standalone test script.

For more details I'll probably have to build and run your code. Let me know if this isn't enough.

xuhdev commented:
@ezyang Thanks. Could you give an example of a minimum code snippet that test a particular operation in JIT, say a simple addition between a tensor and a scalar? I know one can always construct an NN, just wonder whether there's an easier way.

ezyang commented:
I don't know off the top of my head here. I usually read the code mentioned in the backtrace and then make a guess about a stand-alone reproducer that way. It would be nice if we could easily produce standalone reproductions :/

xuhdev commented:
@ezyang OK, I think we might be at cross purpose here. When I look to `File "/var/lib/jenkins/workspace/test/jit_utils.py", line 592`, I got this code snippet:

```python
def enable_cpu_fuser(fn):
    def wrapper(*args, **kwargs):
        torch._C._jit_override_can_fuse_on_cpu(True)
        try:
            fn(*args, **kwargs)
        finally:
            torch._C._jit_override_can_fuse_on_cpu(False)
    return wrapper
```

I searched for `test_add_constant`, and only found it in  test_jit.py:17440, where it is listed as one element of `UBSAN_BLACKLISTED_TESTS`:

```python
UBSAN_BLACKLISTED_TESTS = [
...
    "test_add_constant",
...
]
```

Searching for UBSAN_BLACKLISTED_TESTS also does not lead to any test code, as far as I can tell. I'm guessing that the test means that a constant + a tensor will produce error in JIT by assuming the constant is autograd, but I'm still struggling to reproduce this exact bug...

xuhdev commented:
Perhaps @suo may provide some insight?

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

suo commented:
I'm not sure based on the Python traceback, a C++ level stacktrace would be helpful (basically, if you could run it under gdb and do `bt`)

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev commented:
@suo Thanks, I solved the previous error (and a bunch of other errors) by adding entries to `derivatives.yml`.

Now the only issue that is left seems to be the return type of `add` and `mul` (and maybe `sub` and `div`). For example, `test_binary_op_shape` reports

```
E                   RuntimeError: Expected to find "Float()" but did not find it
E                   7 defined in (%7 : Tensor = aten::mul(%arg1.1, %arg2.1) # <string>:5:11
E                   ~~~~~~~ <--- HERE
E                   )
E                   From CHECK: Float()
```

The code that was compiled is:

```python
def func():
    arg1 = 1
    arg2 = torch.tensor(1, dtype=torch.float32)
    return torch.mul(arg1, arg2)
```

Part of it is compiled to

```python
(%7 : Tensor = aten::mul(%arg1.1, %arg2.1) # <string>:5:11
```

where I think `Tensor` is supposed to be `Float()` in this situation (according to what the test on master says).

My suspicion is that this might be caused by the need to update `symbolic_script.cpp`, but the [instruction to update it](https://github.com/pytorch/pytorch/blob/bcb7c22679315c53390ab7ebfcf02bff5e1a1233/torch/csrc/jit/docs/OVERVIEW.md#test-autodiff) seems outdated: `common_method_tests.py` no longer exists.

Could you help me out on this issue, please? Thanks!

ezyang commented:
cc @suo or perhaps @eellison ?

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev commented:
ping @suo @eellison

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]

ezyang commented:
@xuhdev are you still working on this or do you need another review round?

xuhdev commented:
I'm still working on this (but slowly). I probably need some help, as it seems like a lot of JIT code is involved, which I'm unfamiliar with...

xuhdev commented:
@ezyang I'm unlikely to work on this PR this week (like the past week), and I'm not sure about the week after this week. Do you have an urgent need of this PR? If so, I'm OK with letting someone else take over it.

ezyang commented:
Nope, not urgent, just trying to make sure people aren't blocked.' It looks like you are blocked, but I'm not sure if there's someone on our side with time to help you unblock, sorry :(

xuhdev commented:
@ezyang That's awesome! Let me come back to this some time later.

xuhdev committed:
Update on "Add functions add/sub/rsub/mul/div that accept a scalar as first arg"


Because #29238 removes special processing of `add/sub/rsub/mul/div`, calls like `torch.add(Scalar, Tensor)` will not be automatically translated to `torch.add(Tensor, Tensor)`. Therefore, we explicitly overload these ops with a scalar as their first arg.


[ghstack-poisoned]
